File: scheduler/sync.go
// Copyright 2023 The Yock Authors. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package scheduler

import (
	"sync"

	lua "github.com/yuin/gopher-lua"
	luar "layeh.com/gopher-luar"
)

func loadSync(yocks *YockScheduler) lua.LValue {
	return yocks.registerLib(syncLib)
}

var syncLib = luaFuncs{
	"new": syncNew,
}

// @return userdata
func syncNew(l *lua.LState) int {
	l.Push(luar.New(l, &sync.WaitGroup{}))
	return 1
}

File: lib/include/sync.lua
-- Copyright 2023 The Yock Authors. All rights reserved.
-- Use of this source code is governed by a MIT-style
-- license that can be found in the LICENSE file.

---@meta _

---
---@class waitGroup
---
---{{.wait_group}}
---
local waitGroup = {}

---
---{{.wait_group_add}}
---
---@param delta number
---
function waitGroup:Add(delta)
end

---
---{{.wait_group_done}}
---
function waitGroup:Done()
end

---
---{{.wait_group_wait}}
---
function waitGroup:Wait()
end

---
---{{.sync}}
---
sync = {}

---
---{{.sync_new}}
---
---@return waitGroup
---
function sync.new()
    return {}
end
